### PR TITLE
Fix some undefined cmake variables

### PR DESCRIPTION
**Pull request by jamiesnape**
Toward #3924.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/4003)

### CODE REVIEW

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-clang-ninja-experimental please
@drake-jenkins-bot linux-gcc-experimental please
@drake-jenkins-bot mac-clang-experimental please
@drake-jenkins-bot mac-clang-experimental-matlab please
@drake-jenkins-bot mac-clang-ninja-experimental please
@drake-jenkins-bot mac-clang-ninja-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot mac-clang-experimental-matlab please
@drake-jenkins-bot mac-clang-ninja-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot mac-clang-ninja-experimental-matlab please

**Comment by jamiesnape** [comment]
+@billhoffman for feature review and +@david-german-tri for platform review.

File: drake/bindings/swig/CMakeLists.txt [hunk 1]
```diff
@@ -1,12 +1,14 @@
 list(APPEND CMAKE_MODULE_PATH ${CMAKE_INSTALL_PREFIX}/share/cmake)
 drake_find_package(SWIG MODULE REQUIRED)
-set(DRAKE_SWIG_FILES ${CMAKE_INSTALL_PREFIX}/share/swig)
-include(DrakeUseSWIG)
 
 # Ignore SWIG errors about unknown base types
-set(CMAKE_SWIG_FLAGS ${CMAKE_SWIG_FLAGS} "-w315,-401")
+list(APPEND CMAKE_SWIG_FLAGS "-w315,-401")
 # Ignore SWIG errors about not being able to wrap operator <<
```
Comment (billhoffman) [inline drake/bindings/swig/CMakeLists.txt]: Did you test with cmake --warn-uninitialized after you made this change?

Comment (jamiesnape) [inline drake/bindings/swig/CMakeLists.txt]: Yes.

**Comment by mwoehlke-kitware** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/CMakeLists.txt, line 181 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVe_ig5ww5rqX6f5ec7:-KVe_ig5ww5rqX6f5ec8:b-96uh74) ([raw file](https://github.com/robotlocomotion/drake/blob/42d36d9c935ff7b4abf7824203708ef92215cdb6/drake/CMakeLists.txt#L181)):*
> ```CMake
>       message(STATUS "Found bot2-lcmgl")
>     endif()
>     set(CMAKE_JAVA_INCLUDE_PATH "${LCMGL_JAR_FILE}")
> ```

I don't love that this will override `CMAKE_JAVA_INCLUDE_PATH` if set by the user :disappointed: 

---

*[drake/lcmtypes/CMakeLists.txt, line 69 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVea4OXCByzKJnkEbzz:-KVea4OXCByzKJnkEc--:bmyo95w) ([raw file](https://github.com/robotlocomotion/drake/blob/50d00eba9d8196d9bdb852505f4ce93c49c66be9/drake/lcmtypes/CMakeLists.txt#L69)):*
> ```CMake
>   EXPORT ${PROJECT_NAME}Targets
>   RUNTIME DESTINATION bin
>   LIBRARY DESTINATION lib${LIB_SUFFIX}
> ```

please don't; this change is a regression

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4003)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/lcmtypes/CMakeLists.txt, line 69 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVea4OXCByzKJnkEbzz:-KVeaSIQsEiSyf3rAtE_:b-25s8jy) ([raw file](https://github.com/robotlocomotion/drake/blob/50d00eba9d8196d9bdb852505f4ce93c49c66be9/drake/lcmtypes/CMakeLists.txt#L69)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> please don't; this change is a regression
</details>
Not if CMake does not define it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4003)*

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-clang-ninja-experimental please
@drake-jenkins-bot linux-gcc-experimental please
@drake-jenkins-bot mac-clang-experimental please
@drake-jenkins-bot mac-clang-experimental-matlab please
@drake-jenkins-bot mac-clang-ninja-experimental please
@drake-jenkins-bot mac-clang-ninja-experimental-matlab please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot mac-clang-ninja-experimental please

**Comment by jamiesnape** [comment]
@drake-jenkins-bot mac-clang-experimental please

**Comment by jamiesnape** [comment]
-@billhoffman +@mwoehlke-kitware for feature review.

**Comment by mwoehlke-kitware** [comment]
:lgtm: ...with one non-blocking question

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmake/externals.cmake, line 190 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVfJxowO6OISyHraU4c:-KVfJxoxjJc-yMqmfLSl:b-5dldqz) ([raw file](https://github.com/robotlocomotion/drake/blob/a6150ee0fda19ba2fe07329a35a0db9225aa1b36/cmake/externals.cmake#L190)):*
> ```CMake
>     Java_VERSION_STRING
>     CMAKE_JAVA_COMPILE_FLAGS
>     LIB_SUFFIX)
> ```

TY! (I was thinking yesterday about how this is not propagated and should be, then forgot again; thanks for adding it in this PR!)

---

*[drake/CMakeLists.txt, line 181 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVe_ig5ww5rqX6f5ec7:-KVfMCrduDOUol_n0xm2:bs3r61n) ([raw file](https://github.com/robotlocomotion/drake/blob/42d36d9c935ff7b4abf7824203708ef92215cdb6/drake/CMakeLists.txt#L181)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> I don't love that this will override `CMAKE_JAVA_INCLUDE_PATH` if set by the user :disappointed: 
</details>
...and has been fixed; thanks!

---

*[drake/bindings/swig/CMakeLists.txt, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVfLJQ49tJ9Q8bcn_WO:-KVfLJQ5MCK24mW9tgy0:b76mexq) ([raw file](https://github.com/robotlocomotion/drake/blob/a6150ee0fda19ba2fe07329a35a0db9225aa1b36/drake/bindings/swig/CMakeLists.txt#L9)):*
> ```CMake
> list(APPEND CMAKE_SWIG_FLAGS "-w503")
> 
> set(DRAKE_SWIG_FILES ${CMAKE_INSTALL_PREFIX}/share/swig)
> ```

BTW, why did this need to move?

---

*[drake/lcmtypes/CMakeLists.txt, line 69 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVea4OXCByzKJnkEbzz:-KVfK9Ponbq2JTyDVEt_:b-2vsjij) ([raw file](https://github.com/robotlocomotion/drake/blob/50d00eba9d8196d9bdb852505f4ce93c49c66be9/drake/lcmtypes/CMakeLists.txt#L69)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> Not if CMake does not define it.
</details>
FTR: it would be a regression because removing `LIB_SUFFIX` (and not using instead an alternative such as `LIB_INSTALL_DIR`) here would remove any chance of installing to the correct library directory.

Anyway, this was addressed in the latest revision by making `LIB_SUFFIX` cache-visible so that this does not need to change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-:-KVfMQ4-ElJ7a0YfgPD6:b-bew56s)*

**Comment by jamiesnape** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/swig/CMakeLists.txt, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-KVfLJQ49tJ9Q8bcn_WO:-KVfMxM9Fw6CJvhLzomh:byjtbdd) ([raw file](https://github.com/robotlocomotion/drake/blob/a6150ee0fda19ba2fe07329a35a0db9225aa1b36/drake/bindings/swig/CMakeLists.txt#L9)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary>
> BTW, why did this need to move?
</details>
UseSWIG in DrakeUseSWIG was complaining.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4003)*

**Comment by jamiesnape** [comment]
+@david-german-tri for platform review.

**Comment by jamiesnape** [comment]
Resolved merge conflict.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-clang-ninja-experimental please
@drake-jenkins-bot linux-gcc-experimental please
@drake-jenkins-bot mac-clang-experimental please
@drake-jenkins-bot mac-clang-ninja-experimental please

**Comment by jamiesnape** [comment]
Seems to have broken AVL and XFOIL tests.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-gcc-ninja-experimental-matlab please
@drake-jenkins-bot mac-clang-experimental-matlab please
@drake-jenkins-bot mac-clang-ninja-experimental-matlab please

**Comment by jamiesnape** [comment]
I can only think `CMAKE_EXE_LINKER_FLAGS` are to blame.

**Comment by mwoehlke-kitware** [comment]
`-Wl,--no-undefined` (and its clang cousin) should actually be suprefluous on executables; executables (unlike libraries) don't usually allow lazy linking. At any rate, they should cause relatively obvious build errors, if anything. Not _test failures_ (that aren't due to build errors).

**Comment by jamiesnape** [comment]
Exactly. Basically, I have no idea what is going on with the Fortran executables.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot retest this please

**Comment by jamiesnape** [comment]
Ok. I am going to punt on the last uninitialized variable while I work out how the whole Fortran / MATLAB combo works.

**Comment by jamiesnape** [comment]
-@jamiesnape. +@david-german-tri for platform review.

**Comment by jamiesnape** [comment]
Resolved merge conflicts.

**Comment by jamiesnape** [comment]
@drake-jenkins-bot linux-clang-ninja-experimental please
@drake-jenkins-bot linux-gcc-experimental please

**Comment by david-german-tri** [comment]
:lgtm:

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4003#-:-KVzNjjSI8bVekJGvzkl:bnfp4nl)*

**Comment by david-german-tri** [comment]
Reviewed 2 of 5 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4003)*

**Pull request merged**